### ISSUE TITLE
Grab the content of a QCefWidget to a QImage

### ISSUE THREAD

**Issue by ShelleyLake** [winsoft666/QCefWidget#8]
Dear developer,
I am trying to find a solution to how I can capture the QWidget (QCefWidget) content and save it in a `QImage`.
I tried with the typical "`myWidget->grab()`" function or the equivalent using a `QPainter`, but the result doesn't change:
I only get the background content of the `QWidget `but not the real content (web page).
Do you have any idea how to do this?

Thanks in advance.
_J._

**Comment by winsoft666** [comment]
`QCefWidget->grab()` maybe only work with non-osr mode.

**Comment by ShelleyLake** [comment]
I had no luck at all grabbing the content of the Widget with Qt, neither ON screen nor OFF screen.
I was able to grab the OFF screen widget this way, but it does not work when the content is actually displayed:

```
  QPainter paint(pWidget_);
  CefRenderBuffer* pRenderBuffer = pQCefViewHandler_->lockViewBuffer();
  if (pRenderBuffer) {
    QImage image(pRenderBuffer->buffer.get(),
                 pRenderBuffer->width,
                 pRenderBuffer->height,
                 QImage::Format_ARGB32);
    paint.drawImage(pWidget_->rect(), image, pWidget_->rect());
    result = image.save("c:/image.jpg");
  }
  pQCefViewHandler_->unlockViewBuffer();
```

**Comment by winsoft666** [comment]
what's the meaning of 'but it does not work when the content is actually displayed'?

In OSR mode, you can grab each frame from `QCefBrowserHandler::OnPaint` function.
In Non-OSR mode, you can make an screenshot of CEF render window.

**Comment by ShelleyLake** [comment]
Ah, sorry, I messed up the terms.

Situation 1 - I want to show web content with which the user can interact: in this case I use the non-OSR mode and the window (or better the widget) is visible.
Situation 2 - I would like to load web content, interact with it through javascript, but without showing anything to the user: in this case I use the OSR mode and the widget is hidden.

Then I was looking for a way to capture and put into a QImage the web content (not necessarily saved on disk) regardless of the mode (OSR mode or not).
Since I did not have an idea how to do that, and the QCefWidget->grab() only captured the backgraound of the QWidget itself, i came up with the above code, but it works only for OSR mode.
